pkg/oidc/provider/userinfo.go

```diff
@@ -0,0 +1,125 @@
+	jwks                signingKeyGetter
+}
+
+func newUserInfoHandler(userLister wrangmgmtv3.UserCache, userAttributeLister wrangmgmtv3.UserAttributeCache, jwks signingKeyGetter) *userInfoHandler {
+
+	return &userInfoHandler{
+		userCache:           userLister,
+		userAttributeLister: userAttributeLister,
+		jwks:                jwks,
+	}
+}
+
+// userInfoEndpoint handles the userinfo endpoint of the OIDC provider
+func (h *userInfoHandler) userInfoEndpoint(w http.ResponseWriter, r *http.Request) {
+	accessToken, err := getTokenFromHeader(r)
+	claims := jwt.MapClaims{}
+	// verify access_token signature
+	_, err = jwt.ParseWithClaims(accessToken, &claims, func(token *jwt.Token) (interface{}, error) {
+		// Ensure correct signing method
+		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+		}
+		kid, ok := token.Header["kid"].(string)
+		if !ok {
+			return nil, fmt.Errorf("can't find kid")
+		}
+		pubKey, err := h.jwks.GetPublicKey(kid)
+		if err != nil {
+			return nil, err
+		}
+
+		return pubKey, nil
+	})
+	if err != nil {
+		oidcerror.WriteError(oidcerror.InvalidRequest, fmt.Sprintf("invalid access_token: %v", err), http.StatusBadRequest, w)
+		return
+	}
+
+	userId, ok := claims["sub"].(string)
+	if !ok {
+		oidcerror.WriteError(oidcerror.InvalidRequest, "invalid access_token: it doesn't have sub", http.StatusBadRequest, w)
+		return
+	}
+	scopes, ok := claims["scope"].([]interface{})
+	if !ok {
+		oidcerror.WriteError(oidcerror.InvalidRequest, "invalid access_token: it doesn't have scope", http.StatusBadRequest, w)
+		return
+	}
+	response := UserInfoResponse{
+		Sub: userId,
+	}
+	if slices.Contains(scopes, "profile") {
+		user, err := h.userCache.Get(userId)
+		if err != nil {
+			oidcerror.WriteError(oidcerror.ServerError, fmt.Sprintf("can't get user: %v", err), http.StatusInternalServerError, w)
+			return
+		}
+		response.UserName = user.DisplayName
+	}
+
+	attribs, err := h.userAttributeLister.Get(userId)
+	if err != nil && !apierrors.IsNotFound(err) {
+		oidcerror.WriteError(oidcerror.ServerError, fmt.Sprintf("can't get user attributes: %v", err), http.StatusInternalServerError, w)
+		return
+	}
+	var groups []string
+	if attribs != nil {
+		for _, gps := range attribs.GroupPrincipals {
+			for _, principal := range gps.Items {
+				name := strings.TrimPrefix(principal.Name, "local://")
+				groups = append(groups, name)
+			}
+		}
+	}
+
+	if groups != nil {
+		response.Groups = groups
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(&response); err != nil {
+		http.Error(w, "failed to encode user info response", http.StatusInternalServerError)
+	}
+
+}
+
+func getTokenFromHeader(r *http.Request) (string, error) {
+	authHeader := r.Header.Get("Authorization")
+	if authHeader == "" {
+		return "", fmt.Errorf("authorization header is missing")
+	}
+
+	// Expected format: "Bearer <token>"
```

Comment:
This might be a use for `strings.Cut()`

I think you could do `before, after, found := strings.Cut("Bearer token", " ")`

In which case `before` would be `"Bearer"` and `after` would be `"token"` and `found` would be true.

https://go.dev/play/p/IzEzMPe9aeh

Reply:
yes it looks much better with `strings.Cut()` Thanks!
changed in https://github.com/rancher/rancher/pull/49558/commits/607697782782262b79a64b7cf881cfa1859655d2